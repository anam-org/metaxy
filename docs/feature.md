## Feature Version

Every feature has a **feature version** - hash of its complete specification:
- Feature key
- Field definitions and code versions
- Dependencies (feature-level and field-level)

Being more pedantic, the feature version is calculated by hashing together its field versions and the feature key. The field version is calculated by hashing together the field code version, key, and versions of its dependencies.

Feature and field versions are static and deterministic. They are available as class methods:

```python
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=1),
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

# Get feature version
print(VideoProcessing.feature_version())  # "a3f8b2c1"

# Change code_version
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=2),  # Changed!
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

print(VideoProcessing.feature_version())  # "d7e9f4a2" (different!)
```

Every metadata row includes a `feature_version` column:
```python
metadata = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "data_version": [...],
    "feature_version": "a3f8b2c1",  # Auto-added by store
})
```

This enables:
- **Current vs historical**: Filter by `feature_version` to get specific versions
- **Migration tracking**: Know which rows need updating
- **Audit trail**: See what feature definition produced each row


## User-Defined Metadata Alignment

When feature definitions change, you may need custom logic to align metadata with upstream changes. Override `Feature.align_metadata_with_upstream`:

### Default Behavior (Inner Join)

```python
class MyFeature(Feature, spec=...):
    # Default behavior: inner join on sample_id
    # Only keeps samples present in ALL upstream features
    pass

# During migration:
# - Reads current metadata
# - Inner joins with upstream on sample_id
# - Only samples present in both current AND upstream survive
```

### Custom Alignment Examples

#### One-to-Many Mapping

```python
class VideoFrames(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "frames"]),
    deps=[FeatureDep(key=FeatureKey(["videos"]))],
    ...
)):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Each video produces 30 frames.

        When videos upstream change, regenerate all frame sample IDs.
        """
        video_samples = upstream_metadata["videos"]["sample_id"]

        # Generate frame sample IDs
        frames = []
        for video_id in video_samples:
            for frame_idx in range(30):
                frames.append({
                    "sample_id": f"{video_id}_frame_{frame_idx}",
                    "video_id": video_id,
                    "frame_idx": frame_idx,
                })

        return pl.DataFrame(frames)
```

#### Conditional Filtering

```python
class ProcessedVideos(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Only process videos longer than 10 seconds."""

        videos = upstream_metadata["videos"]

        # Filter by duration
        valid_videos = videos.filter(pl.col("duration") > 10)

        # Keep custom columns from current_metadata if available
        if len(current_metadata) > 0:
            return current_metadata.join(
                valid_videos.select(pl.col("sample_id")),
                on="sample_id",
                how="inner",  # Only keep samples that pass filter
            )
        else:
            return valid_videos.select(["sample_id", "duration", "path"])
```

#### Outer Join (Keep All Samples)

```python
class MergedFeature(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Keep union of all upstream samples (outer join)."""

        all_samples = set()
        for upstream_df in upstream_metadata.values():
            all_samples.update(upstream_df["sample_id"].to_list())

        return pl.DataFrame({"sample_id": sorted(all_samples)})
```
