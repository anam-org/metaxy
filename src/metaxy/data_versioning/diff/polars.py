"""Polars implementation of metadata diff resolver."""

import polars as pl

from metaxy.data_versioning.diff.base import DiffResult, MetadataDiffResolver


class PolarsDiffResolver(MetadataDiffResolver):
    """Identifies changed rows using Polars operations.

    Strategy:
    - Categorizes changes into added, changed, and removed
    - Uses LEFT/RIGHT JOINs to identify each category
    - Materializes once and splits into three DataFrames (efficient)
    """

    def find_changes(
        self,
        target_versions: pl.LazyFrame,
        current_metadata: pl.LazyFrame | None,
    ) -> DiffResult:
        """Find all changes between target and current.

        Args:
            target_versions: LazyFrame with calculated data_versions
            current_metadata: LazyFrame with current metadata, or None

        Returns:
            DiffResult with three materialized DataFrames
        """
        # Select only sample_id and data_version from target_versions
        # (it may have intermediate joined columns from upstream)
        target_versions = target_versions.select(["sample_id", "data_version"])

        if current_metadata is None:
            # No existing metadata - all target rows are new, nothing removed/changed
            empty_df = pl.DataFrame({"sample_id": [], "data_version": []})
            return DiffResult(
                added=target_versions.collect(),
                changed=empty_df,
                removed=empty_df,
            )

        # Join target with current to categorize changes
        # Keep only sample_id and data_version from current for comparison
        current_comparison = current_metadata.select(
            ["sample_id", pl.col("data_version").alias("__current_data_version")]
        )

        # LEFT JOIN target with current
        compared = target_versions.join(
            current_comparison,
            on="sample_id",
            how="left",
        )

        # Build lazy queries for each category
        added_lazy = (
            compared.filter(pl.col("__current_data_version").is_null())
            .drop("__current_data_version")
            .select(["sample_id", "data_version"])
        )

        changed_lazy = (
            compared.filter(
                pl.col("__current_data_version").is_not_null()
                & (pl.col("data_version") != pl.col("__current_data_version"))
            )
            .drop("__current_data_version")
            .select(["sample_id", "data_version"])
        )

        removed_lazy = current_metadata.join(
            target_versions.select("sample_id"),
            on="sample_id",
            how="anti",
        ).select(["sample_id", "data_version"])

        # Collect all three at once (single execution for shared computations)
        added_df, changed_df, removed_df = pl.collect_all(
            [added_lazy, changed_lazy, removed_lazy]
        )

        return DiffResult(
            added=added_df,
            changed=changed_df,
            removed=removed_df,
        )
