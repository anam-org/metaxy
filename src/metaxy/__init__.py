from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
