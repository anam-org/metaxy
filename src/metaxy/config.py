"""Configuration system for Metaxy using pydantic-settings."""

from pathlib import Path
from typing import TYPE_CHECKING, Any, TypeVar, overload

try:
    import tomllib  # Python 3.11+
except ImportError:
    import tomli as tomllib  # Fallback for Python 3.10

from pydantic import Field as PydanticField
from pydantic_settings import (
    BaseSettings,
    PydanticBaseSettingsSource,
    SettingsConfigDict,
)

if TYPE_CHECKING:
    from metaxy.metadata_store.base import MetadataStore
    from metaxy.models.feature import FeatureRegistry

T = TypeVar("T")


def load_registry(registry_path: str) -> "FeatureRegistry":
    """Load registry from import path.

    Args:
        registry_path: Import path like "myapp.features:my_registry"

    Returns:
        FeatureRegistry instance

    Raises:
        ValueError: If path is invalid
        ImportError: If module or registry not found
        TypeError: If loaded object is not a FeatureRegistry
    """
    from metaxy.models.feature import FeatureRegistry

    if ":" not in registry_path:
        raise ValueError(
            f"Invalid registry path: {registry_path}. "
            f"Expected format: 'module.path:registry_name'"
        )

    module_path, registry_name = registry_path.split(":", 1)

    try:
        module = __import__(module_path, fromlist=[registry_name])
        registry = getattr(module, registry_name)
    except (ImportError, AttributeError) as e:
        raise ImportError(f"Failed to load registry from {registry_path}: {e}") from e

    if not isinstance(registry, FeatureRegistry):
        raise TypeError(
            f"Registry at {registry_path} is not a FeatureRegistry instance. "
            f"Got: {type(registry)}"
        )

    return registry


class TomlConfigSettingsSource(PydanticBaseSettingsSource):
    """Custom settings source for TOML configuration files.

    Auto-discovers configuration in this order:
    1. Explicit file path if provided
    2. metaxy.toml in current directory (preferred)
    3. pyproject.toml [tool.metaxy] section (fallback)
    4. No config (returns empty dict)
    """

    def __init__(self, settings_cls: type[BaseSettings], toml_file: Path | None = None):
        super().__init__(settings_cls)
        self.toml_file = toml_file or self._discover_config_file()
        self.toml_data = self._load_toml()

    def _discover_config_file(self) -> Path | None:
        """Auto-discover config file."""
        # Prefer metaxy.toml
        if Path("metaxy.toml").exists():
            return Path("metaxy.toml")

        # Fallback to pyproject.toml
        if Path("pyproject.toml").exists():
            return Path("pyproject.toml")

        return None

    def _load_toml(self) -> dict[str, Any]:
        """Load TOML file and extract metaxy config."""
        if self.toml_file is None:
            return {}

        with open(self.toml_file, "rb") as f:
            data = tomllib.load(f)

        # Extract [tool.metaxy] from pyproject.toml or root from metaxy.toml
        if self.toml_file.name == "pyproject.toml":
            return data.get("tool", {}).get("metaxy", {})
        else:
            return data

    def get_field_value(self, field: Any, field_name: str) -> tuple[Any, str, bool]:
        """Get field value from TOML data."""
        field_value = self.toml_data.get(field_name)
        return field_value, field_name, False

    def __call__(self) -> dict[str, Any]:
        """Return all settings from TOML."""
        return self.toml_data


class StoreConfig(BaseSettings):
    """Configuration for a single metadata store.

    Structure:
        type: Full import path to store class
        config: Dict of all configuration (including fallback_stores)

    Example:
        >>> config = StoreConfig(
        ...     type="metaxy_delta.DeltaMetadataStore",
        ...     config={
        ...         "table_uri": "s3://bucket/metadata",
        ...         "region": "us-west-2",
        ...         "fallback_stores": ["prod"],
        ...     }
        ... )
    """

    model_config = SettingsConfigDict(
        extra="forbid",  # Only type and config fields allowed
    )

    # Store class (full import path)
    type: str

    # Store configuration (all kwargs for __init__)
    # This includes fallback_stores, table_uri, db_path, storage_options, etc.
    config: dict[str, Any] = PydanticField(default_factory=dict)


class MetaxyConfig(BaseSettings):
    """Main Metaxy configuration.

    Loads from:
    1. TOML file (metaxy.toml or pyproject.toml [tool.metaxy])
    2. Environment variables (METAXY_*)
    3. Init arguments

    Priority: init > env vars > TOML

    Example:
        >>> # Auto-discover config
        >>> config = MetaxyConfig.load()
        >>>
        >>> # Get store instance
        >>> store = config.get_store("prod")
        >>>
        >>> # Override via env var
        >>> # METAXY_STORE=staging METAXY_REGISTRY=myapp.features:my_registry
        >>> config = MetaxyConfig.load()
        >>> store = config.get_store()  # Uses staging with custom registry
    """

    model_config = SettingsConfigDict(
        env_prefix="METAXY_",
        env_nested_delimiter="__",
    )

    # Store to use
    store: str = "dev"

    # Named store configurations
    stores: dict[str, StoreConfig] = PydanticField(default_factory=dict)

    # Migrations directory
    migrations_dir: str = "metaxy/migrations"

    # Entrypoints to load (list of module paths)
    entrypoints: list[str] = PydanticField(default_factory=list)

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> tuple[PydanticBaseSettingsSource, ...]:
        """Customize settings sources: init → env → TOML.

        Priority (first wins):
        1. Init arguments
        2. Environment variables
        3. TOML file
        """
        toml_settings = TomlConfigSettingsSource(settings_cls)
        return (init_settings, env_settings, toml_settings)

    @classmethod
    def load(
        cls, config_file: str | Path | None = None, *, search_parents: bool = True
    ) -> "MetaxyConfig":
        """Load config with auto-discovery and parent directory search.

        Args:
            config_file: Optional config file path (overrides auto-discovery)
            search_parents: Search parent directories for config file (default: True)

        Returns:
            Loaded config (TOML + env vars merged)

        Example:
            >>> # Auto-discover with parent search
            >>> config = MetaxyConfig.load()
            >>>
            >>> # Explicit file
            >>> config = MetaxyConfig.load("custom.toml")

            >>> # Auto-discover without parent search
            >>> config = MetaxyConfig.load(search_parents=False)
        """
        # Search for config file if not explicitly provided
        if config_file is None and search_parents:
            config_file = cls._discover_config_with_parents()

        # For explicit file, temporarily patch the TomlConfigSettingsSource
        # to use that file, then use normal instantiation
        # This ensures env vars still work

        if config_file:
            # Create a custom settings source class for this file
            toml_path = Path(config_file)

            class CustomTomlSource(TomlConfigSettingsSource):
                def __init__(self, settings_cls: type[BaseSettings]):
                    # Skip auto-discovery, use explicit file
                    super(TomlConfigSettingsSource, self).__init__(settings_cls)
                    self.toml_file = toml_path
                    self.toml_data = self._load_toml()

            # Customize sources to use custom TOML file
            original_method = cls.settings_customise_sources

            @classmethod  # type: ignore[misc]
            def custom_sources(
                cls_inner,
                settings_cls,
                init_settings,
                env_settings,
                dotenv_settings,
                file_secret_settings,
            ):
                toml_settings = CustomTomlSource(settings_cls)
                return (init_settings, env_settings, toml_settings)

            # Temporarily replace method
            cls.settings_customise_sources = custom_sources  # type: ignore[assignment]
            config = cls()
            cls.settings_customise_sources = original_method  # type: ignore[method-assign]
        else:
            # Use default sources (auto-discovery + env vars)
            config = cls()

        # Load entrypoints if configured (do this after config is created)
        if config.entrypoints:
            from metaxy.entrypoints import load_config_entrypoints

            load_config_entrypoints(config.entrypoints)

        return config

    @staticmethod
    def _discover_config_with_parents() -> Path | None:
        """Discover config file by searching current and parent directories.

        Searches for metaxy.toml or pyproject.toml in current directory,
        then iteratively searches parent directories.

        Returns:
            Path to config file if found, None otherwise
        """
        current = Path.cwd()

        while True:
            # Check for metaxy.toml (preferred)
            metaxy_toml = current / "metaxy.toml"
            if metaxy_toml.exists():
                return metaxy_toml

            # Check for pyproject.toml
            pyproject_toml = current / "pyproject.toml"
            if pyproject_toml.exists():
                return pyproject_toml

            # Move to parent
            parent = current.parent
            if parent == current:
                # Reached root
                break
            current = parent

        return None

    def get_store(
        self,
        name: str | None = None,
    ) -> "MetadataStore":
        """Instantiate metadata store by name.

        Args:
            name: Store name (uses config.store if None)

        Returns:
            Instantiated metadata store

        Raises:
            ValueError: If store name not found in config, or if fallback stores
                have different hash algorithms than the parent store
            ImportError: If store class cannot be imported

        Example:
            >>> config = MetaxyConfig.load()
            >>> store = config.get_store("prod")
            >>>
            >>> # Use default store
            >>> store = config.get_store()
        """
        from metaxy.data_versioning.hash_algorithms import HashAlgorithm

        if len(self.stores) == 0:
            raise ValueError(
                "No Metaxy stores available. They should be configured in metaxy.toml|pyproject.toml or via environment variables."
            )

        name = name or self.store

        if name not in self.stores:
            raise ValueError(
                f"Store '{name}' not found in config. "
                f"Available stores: {list(self.stores.keys())}"
            )

        store_config = self.stores[name]

        # Import store class
        store_class = self._import_class(store_config.type)

        # Extract configuration
        config_copy = store_config.config.copy()
        fallback_store_names = config_copy.pop("fallback_stores", [])

        # Get hash_algorithm from config (if specified) and convert to enum
        configured_hash_algorithm = config_copy.get("hash_algorithm")
        if configured_hash_algorithm is not None:
            # Convert string to enum if needed
            if isinstance(configured_hash_algorithm, str):
                configured_hash_algorithm = HashAlgorithm(configured_hash_algorithm)
                config_copy["hash_algorithm"] = configured_hash_algorithm
        else:
            # Use default
            configured_hash_algorithm = HashAlgorithm.XXHASH64
            config_copy["hash_algorithm"] = configured_hash_algorithm

        # Build fallback stores recursively
        fallback_stores = []
        for fallback_name in fallback_store_names:
            fallback_store = self.get_store(fallback_name)
            fallback_stores.append(fallback_store)

        # Instantiate store with config + fallback_stores
        store = store_class(
            fallback_stores=fallback_stores,
            **config_copy,
        )

        # Verify the store actually uses the hash algorithm we configured
        # (in case a store subclass overrides the default or ignores the parameter)
        if store.hash_algorithm != configured_hash_algorithm:
            raise ValueError(
                f"Store '{name}' ({store_class.__name__}) was configured with "
                f"hash_algorithm='{configured_hash_algorithm.value}' but is using "
                f"'{store.hash_algorithm.value}'. The store class may have overridden "
                f"the hash algorithm. All stores must use the same hash algorithm."
            )

        return store

    @staticmethod
    def _import_class(class_path: str) -> type:
        """Import class from module path.

        Args:
            class_path: Full import path like "metaxy.metadata_store.InMemoryMetadataStore"

        Returns:
            Imported class

        Raises:
            ImportError: If module or class not found
        """
        module_path, class_name = class_path.rsplit(".", 1)
        module = __import__(module_path, fromlist=[class_name])
        return getattr(module, class_name)

    @staticmethod
    @overload
    def _import_object(import_path: str, expected_type: type[T]) -> T: ...

    @staticmethod
    @overload
    def _import_object(import_path: str, expected_type: None = None) -> Any: ...

    @staticmethod
    def _import_object(
        import_path: str, expected_type: type[T] | None = None
    ) -> T | Any:
        """Import object from module:object path.

        Args:
            import_path: Import path like "module.path:object_name"
            expected_type: Optional type to validate against

        Returns:
            Imported object (typed as expected_type if provided)

        Raises:
            ValueError: If path format is invalid
            ImportError: If module or object not found
            TypeError: If object doesn't match expected_type
        """
        if ":" not in import_path:
            raise ValueError(
                f"Invalid import path: {import_path}. "
                f"Expected format: 'module.path:object_name'"
            )

        module_path, object_name = import_path.split(":", 1)

        try:
            module = __import__(module_path, fromlist=[object_name])
            obj = getattr(module, object_name)
        except (ImportError, AttributeError) as e:
            raise ImportError(f"Failed to load object from {import_path}: {e}") from e

        if expected_type is not None and not isinstance(obj, expected_type):
            type_name = getattr(expected_type, "__name__", str(expected_type))
            raise TypeError(
                f"Object at {import_path} is not a {type_name} instance. "
                f"Got: {type(obj)}"
            )

        return obj
