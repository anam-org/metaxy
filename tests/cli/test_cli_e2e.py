"""End-to-end CLI tests with real workflows."""

import subprocess
import sys
from pathlib import Path

import polars as pl
import pytest
from syrupy.assertion import SnapshotAssertion


@pytest.fixture
def e2e_project(tmp_path: Path):
    """Create a complete test project with features, config, and migrations."""
    project_dir = tmp_path / "test_project"
    project_dir.mkdir()

    # Create features module (v1)
    features_dir = project_dir / "features"
    features_dir.mkdir()

    (features_dir / "__init__.py").write_text("")

    (features_dir / "video.py").write_text("""
from metaxy import Feature, FeatureSpec, FeatureKey, FieldSpec, FieldKey

class VideoFiles(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "files"]),
    deps=None,
    fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)]
)):
    pass
""")

    (features_dir / "processing.py").write_text("""
from metaxy import (
    Feature, FeatureSpec, FeatureKey, FieldSpec, FieldKey,
    FeatureDep, FieldDep
)

class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    deps=[FeatureDep(key=FeatureKey(["video", "files"]))],
    fields=[FieldSpec(
        key=FieldKey(["frames"]),
        code_version=1,
        deps=[FieldDep(
            feature_key=FeatureKey(["video", "files"]),
            fields=[FieldKey(["default"])]
        )]
    )]
)):
    pass
""")

    # Create config file with entrypoints and unique database path
    db_path = project_dir / "metadata.duckdb"
    (project_dir / "metaxy.toml").write_text(f"""
store = "dev"
migrations_dir = "migrations"
entrypoints = ["features.video", "features.processing"]

[stores.dev]
type = "metaxy.metadata_store.duckdb.DuckDBMetadataStore"

[stores.dev.config]
database = "{db_path}"
""")

    # Add project to path
    sys.path.insert(0, str(project_dir))

    yield project_dir

    # Cleanup
    if str(project_dir) in sys.path:
        sys.path.remove(str(project_dir))


def test_cli_e2e_duckdb_workflow(e2e_project: Path, snapshot: SnapshotAssertion):
    """Test complete CLI workflow with DuckDB store.

    Workflow:
    1. Load features v1
    2. Write initial metadata
    3. Push (record feature versions)
    4. Update feature code (v2)
    5. Generate migration
    6. Apply migration
    7. Verify results
    """
    pytest.importorskip("duckdb")

    # Change to project directory
    import os

    original_cwd = os.getcwd()
    os.chdir(e2e_project)

    # Ensure project is in path for imports
    sys.path.insert(0, str(e2e_project))

    try:
        # Step 1: Load features and write initial data
        # Use config to load entrypoints (this is what the CLI will do)
        from metaxy import FeatureKey
        from metaxy.config import MetaxyConfig

        # Load config which will auto-load entrypoints
        config = MetaxyConfig.load()

        # Debug: check config
        print(f"Config entrypoints: {config.entrypoints}")
        print(f"Config store: {config.store}")

        # Get features from the now-populated global graph
        from metaxy.models.feature import FeatureGraph

        graph = FeatureGraph.get_active()

        # Debug: check what's loaded
        print(f"Loaded features: {list(graph.features_by_key.keys())}")

        VideoFiles = graph.features_by_key[FeatureKey(["video", "files"])]
        VideoProcessing = graph.features_by_key[FeatureKey(["video", "processing"])]

        # Get store and write initial data
        store = config.get_store()

        with store:
            # Write initial metadata
            video_files_data = pl.DataFrame(
                {
                    "sample_id": ["video1.mp4", "video2.mp4"],
                    "path": ["/data/video1.mp4", "/data/video2.mp4"],
                    "data_version": [{"default": "hash1"}, {"default": "hash2"}],
                }
            )
            store.write_metadata(VideoFiles, video_files_data)

            # Write processing metadata
            processing_data = pl.DataFrame(
                {
                    "sample_id": ["video1.mp4", "video2.mp4"],
                    "data_version": [{"frames": "proc1"}, {"frames": "proc2"}],
                }
            )
            store.write_metadata(VideoProcessing, processing_data)

            # Step 2: Push (record feature versions)
            snapshot_id_v1 = store.serialize_feature_graph()

            assert snapshot_id_v1 is not None
            assert len(snapshot_id_v1) == 64  # Full SHA256 hash

        # Store is now closed, data is persisted to DuckDB

        # Step 3: Update feature code (simulate code change)
        (e2e_project / "features" / "processing.py").write_text("""
from metaxy import (
    Feature, FeatureSpec, FeatureKey, FieldSpec, FieldKey,
    FeatureDep, FieldDep
)

class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    deps=[FeatureDep(key=FeatureKey(["video", "files"]))],
    fields=[FieldSpec(
        key=FieldKey(["frames"]),
        code_version=2,  # Changed from 1 to 2
        deps=[FieldDep(
            feature_key=FeatureKey(["video", "files"]),
            fields=[FieldKey(["default"])]
        )]
    )]
)):
    pass
""")

        # Step 4: Generate migration using CLI subprocess
        # The subprocess will load the new v2 code and compare with v1 in DB
        result = subprocess.run(
            [sys.executable, "-m", "metaxy.cli.app", "migrations", "generate"],
            cwd=str(e2e_project),
            capture_output=True,
            text=True,
        )

        # Debug output if failed
        if result.returncode != 0 or "No feature changes detected" in result.stdout:
            print(f"STDOUT: {result.stdout}")
            print(f"STDERR: {result.stderr}")

        assert result.returncode == 0, f"Generate failed: {result.stderr}"
        assert "No feature changes detected" not in result.stdout, (
            f"No changes detected. Output: {result.stdout}"
        )

        # Verify migration file created (in metaxy/migrations dir from config)
        migrations_dir = e2e_project / config.migrations_dir
        assert migrations_dir.exists(), (
            f"Migrations dir not created. Output: {result.stdout}"
        )

        migration_files = list(migrations_dir.glob("*.yaml"))
        assert len(migration_files) == 1

        migration_file = migration_files[0]

        # Load and verify migration
        from metaxy.migrations import Migration

        migration = Migration.from_yaml(str(migration_file))

        # Should have 1 operation (only processing changed, it has upstream so it's reconcilable)
        ops = migration.get_operations()
        assert len(ops) == 1
        assert ops[0].feature_key == ["video", "processing"]

        # Snapshot the generated migration structure
        migration_snapshot = {
            "version": migration.version,
            "parent_migration_id": migration.parent_migration_id,
            "operations": [
                {
                    "id": op.id,
                    "type": op.type,  # type: ignore[attr-defined]
                    "feature_key": op.feature_key,
                    "reason": op.reason,
                }
                for op in ops
            ],
        }
        assert migration_snapshot == snapshot

        # Step 5: Apply migration using CLI (dry run first)
        # Use migration ID instead of file path
        result = subprocess.run(
            [
                sys.executable,
                "-m",
                "metaxy.cli.app",
                "migrations",
                "apply",
                migration.id,  # Migration ID, not file path
                "--dry-run",
            ],
            cwd=str(e2e_project),
            capture_output=True,
            text=True,
        )

        assert result.returncode == 0, f"Apply dry-run failed: {result.stderr}"
        assert (
            "skipped" in result.stdout.lower() or "completed" in result.stdout.lower()
        )

        # Step 6: Apply migration for real
        result = subprocess.run(
            [
                sys.executable,
                "-m",
                "metaxy.cli.app",
                "migrations",
                "apply",
                migration.id,  # Migration ID, not file path
            ],
            cwd=str(e2e_project),
            capture_output=True,
            text=True,
        )

        assert result.returncode == 0, f"Apply failed: {result.stderr}"

        # Step 7: Verify migration was applied
        # Reload config and graph to get v2 features
        config = MetaxyConfig.load()
        graph = FeatureGraph.get_active()
        VideoProcessingV2 = graph.features_by_key[FeatureKey(["video", "processing"])]

        store = config.get_store()
        with store:
            # Check migration system tables
            from metaxy.migrations.executor import MIGRATIONS_KEY

            migrations = store.read_metadata(MIGRATIONS_KEY, current_only=False)
            assert len(migrations) == 1
            assert migrations["migration_id"][0] == migration.id

            # Verify processing feature was updated
            all_processing = store.read_metadata(
                VideoProcessingV2,
                current_only=False,
            )

            # Should have original + migrated rows
            assert len(all_processing) >= 2

            # Snapshot the final data state
            final_snapshot = {
                "row_count": len(all_processing),
                "sample_ids": sorted(all_processing["sample_id"].unique().to_list()),
                "feature_versions": sorted(
                    all_processing["feature_version"].unique().to_list()
                ),
            }
            assert final_snapshot == snapshot

        # Store is now closed, can run CLI commands without lock conflicts

        # Check migration status via CLI
        status_result = subprocess.run(
            [sys.executable, "-m", "metaxy.cli.app", "migrations", "status"],
            cwd=str(e2e_project),
            capture_output=True,
            text=True,
        )

        assert status_result.returncode == 0, f"Status failed: {status_result.stderr}"

        # Snapshot migration status output (excluding timestamp-based migration_id)
        status_snapshot = {
            "has_migrations": "migration_" in status_result.stdout,
            "migration_completed": "COMPLETED" in status_result.stdout,
            "has_migration_id": migration.id is not None if migration else False,
        }
        assert status_snapshot == snapshot

    finally:
        os.chdir(original_cwd)
        if str(e2e_project) in sys.path:
            sys.path.remove(str(e2e_project))


def test_cli_migration_status_command(e2e_project: Path):
    """Test migrations status command shows correct info."""
    pytest.importorskip("duckdb")

    import os

    original_cwd = os.getcwd()
    os.chdir(e2e_project)
    sys.path.insert(0, str(e2e_project))

    try:
        # First generate a migration (sets up the workflow)
        from metaxy import FeatureKey
        from metaxy.config import MetaxyConfig
        from metaxy.models.feature import FeatureGraph

        # Load config (auto-loads entrypoints)
        config = MetaxyConfig.load()
        graph = FeatureGraph.get_active()

        store = config.get_store()

        with store:
            # Write some data and record versions
            VideoFiles = graph.features_by_key[FeatureKey(["video", "files"])]

            store.write_metadata(
                VideoFiles,
                pl.DataFrame(
                    {
                        "sample_id": ["v1"],
                        "data_version": [{"default": "h1"}],
                    }
                ),
            )
            store.serialize_feature_graph()

        # Update code to trigger a change
        (e2e_project / "features" / "video.py").write_text("""
from metaxy import Feature, FeatureSpec, FeatureKey, FieldSpec, FieldKey

class VideoFiles(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "files"]),
    deps=None,
    fields=[FieldSpec(key=FieldKey(["default"]), code_version=2)]  # Changed!
)):
    pass
""")

        # Check status (should show no migrations initially)
        result = subprocess.run(
            [sys.executable, "-m", "metaxy.cli.app", "migrations", "status"],
            cwd=str(e2e_project),
            capture_output=True,
            text=True,
        )

        assert result.returncode == 0
        assert (
            "No migrations found" in result.stdout
            or "Migration Status:" in result.stdout
        )

    finally:
        os.chdir(original_cwd)
