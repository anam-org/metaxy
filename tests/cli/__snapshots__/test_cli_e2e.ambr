# serializer version: 1
# name: test_cli_e2e_duckdb_workflow
  dict({
    'operations': list([
      dict({
        'feature_key': list([
          'video',
          'processing',
        ]),
        'id': 'reconcile_video/processing',
        'reason': 'TODO: Describe what changed and why',
        'type': 'metaxy.migrations.ops.DataVersionReconciliation',
      }),
    ]),
    'parent_migration_id': None,
    'version': 1,
  })
# ---
# name: test_cli_e2e_duckdb_workflow.1
  dict({
    'feature_versions': list([
      '367ec6fb70be09de7dd481d46be8b1c69a0544ab686acd582a9819255eb5be5e',
      '3c5b4c900fdf1b7f5c08736233f70738a9ca73f94c6aa2b60aeefb5b46105085',
    ]),
    'row_count': 4,
    'sample_ids': list([
      'video1.mp4',
      'video2.mp4',
    ]),
  })
# ---
# name: test_cli_e2e_duckdb_workflow.2
  dict({
    'has_migration_id': True,
    'has_migrations': True,
    'migration_completed': True,
  })
# ---
